safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,182 @@
+)
+
+const (
+	ValueHTTPS         = "https:"
+	ValueHTTP          = "http:"
+	ValueUnsafeEval    = "'unsafe-eval'"
+	ValueUnsafeInline  = "'unsafe-inline'"
+	ValueNone          = "'none'"
+	ValueStrictDynamic = "'strict-dynamic'"
+)
+
+// PolicyDirective contains a single CSP directive.
+type PolicyDirective struct {
+	dir       Directive
+	vals      []string
+	nonces    bool
+	lastNonce string
+	// readRand is used for dependency injection in tests.
+	readRand func([]byte) (int, error)
+}
+
+// NewPolicyDirective creates a new PolicyDirective given the directive, the
+// values and whether or not a nonce directive should be added.
+func NewPolicyDirective(d Directive, values []string, nonces bool) *PolicyDirective {
+	return &PolicyDirective{
+		dir:      d,
+		vals:     values,
+		nonces:   nonces,
+		readRand: rand.Read,
+	}
+}
+
+// nonceSize is the size of the nonces in bytes.
+const nonceSize = 8
+
+func (pd *PolicyDirective) generateNonce() {
+	b := make([]byte, nonceSize)
+	_, err := pd.readRand(b)
+	if err != nil {
+		// TODO: handle this better, what should happen here?
+		panic(err)
+	}
+	pd.lastNonce = base64.StdEncoding.EncodeToString(b)
+}
+
+// Policy defines a CSP policy, containing many directives.
+type Policy struct {
+	Directives []*PolicyDirective
+}
+
+// NewPolicy creates a new strict, nonce-based CSP.
+// See https://csp.withgoogle.com/docs/strict-csp.html for more info.
+//
+// TODO: maybe reportURI should be safehttp.URL?
+func NewPolicy(reportURI string) *Policy {
+	return &Policy{
+		Directives: []*PolicyDirective{
+			NewPolicyDirective(DirectiveObjectSrc, []string{ValueNone}, false),
+			NewPolicyDirective(DirectiveScriptSrc, []string{
+				ValueUnsafeInline,
+				ValueUnsafeEval,
+				ValueStrictDynamic,
+				ValueHTTPS,
+				ValueHTTP,
+			}, true),
+			NewPolicyDirective(DirectiveBaseURI, []string{ValueNone}, false),
+			NewPolicyDirective(DirectiveReportURI, []string{reportURI}, false),
+		},
+	}
+}
+
+// Serialize serializes this policy for use in a Content-Security-Policy header
+// or in a Content-Security-Policy-Report-Only header. The nonces generated for
+// each directive is also returned.
+func (p Policy) Serialize() (csp string, nonces map[Directive]string) {
+	nonces = make(map[Directive]string)
+	b := strings.Builder{}
+
+	for i, d := range p.Directives {
+		if i != 0 {
+			b.WriteString("; ")
+		}
+		b.WriteString(string(d.dir))
+
+		if d.nonces {
+			d.generateNonce()
+			b.WriteString(" 'nonce-")
+			b.WriteString(d.lastNonce)
+			b.WriteString("'")
+			nonces[d.dir] = d.lastNonce
+		}
+
+		for _, v := range d.vals {
+			b.WriteString(" ")
+			b.WriteString(v)
+		}
+	}
+	return b.String(), nonces
+}
+
+// Interceptor intercepts requests and applies CSP policies.
+type Interceptor struct {
+	// EnforcementPolicy will be applied as the Content-Security-Policy header.
+	EnforcementPolicy *Policy
+
+	// ReportOnlyPolicy will be applied as the
+	// Content-Security-Policy-Report-Only header.
```

Comment:
nit: excessive line breaks

Reply:
Well, you see, `____// ReportOnlyPolicy will be applied as the Content-Security-Policy-Report-Only` is wider than 80 characters, so I put a linebreak to follow your earlier code reviews :p

Reply:
Let me clarify:
- very-long lines hurt readability
- excessive line breaking hurts readability.

When there's a very long line, I'll ask you to wrap it, and then it seems natural to set some arbitrary number, be it either 80 or 100 or 120. But there is no hard rule regarding wrapping to X characters.

Note that this is true for Go. I'll usually ask you to wrap e.g. Markdown to 80 chars.

Reply:
I see :+1: